File: built-in plugins dsp/mb-transient-shaper.cpp
/**
 * MB Transient Shaper
 * Category : effect
 * Type     : compressor
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Attack and sustain control for drums and percussive sounds
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_TRANSIENT_SHAPER_H
#define MB_TRANSIENT_SHAPER_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbTransientShaper : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-transient-shaper";
    static constexpr const char* PLUGIN_NAME    = "MB Transient Shaper";
    static constexpr const char* PLUGIN_TYPE    = "compressor";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float attack = 0f;  // range [-100, 100]
    float sustain = 0f;  // range [-100, 100]
    float attack_time = 5f;  // range [0.1, 50]
    float sustain_time = 100f;  // range [10, 500]
    float sensitivity = 0.5f;  // range [0, 1]
    float output = 0f;  // range [-12, 12]
    float mix = 1f;  // range [0, 1]
    };

    MbTransientShaper() = default;
    ~MbTransientShaper() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.attack = std::clamp(params.attack, -100f, 100f);
        params.sustain = std::clamp(params.sustain, -100f, 100f);
        params.attack_time = std::clamp(params.attack_time, 0.1f, 50f);
        params.sustain_time = std::clamp(params.sustain_time, 10f, 500f);
        params.sensitivity = std::clamp(params.sensitivity, 0f, 1f);
        params.output = std::clamp(params.output, -12f, 12f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Transient Shaper
        return input;
    }
};

#endif // MB_TRANSIENT_SHAPER_H

File: built-in plugins dsp/mb-transient-shaper.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbTransientShaperPlugin: PluginDefinition = {
    id: 'mb-transient-shaper',
    slug: 'mb-transient-shaper',
    name: 'MB Transient Shaper',
    category: 'effect',
    type: 'compressor',
    version: '1.0.0',
    description: 'Attack and sustain control for drums and percussive sounds',
    author: 'Max Booster',
    parameters: [
      { id: 'attack', name: 'Attack', type: 'float', defaultValue: 0, minValue: -100, maxValue: 100, unit: '%', automatable: true },
      { id: 'sustain', name: 'Sustain', type: 'float', defaultValue: 0, minValue: -100, maxValue: 100, unit: '%', automatable: true },
      { id: 'attack_time', name: 'Attack Time', type: 'float', defaultValue: 5, minValue: 0.1, maxValue: 50, unit: 'ms', automatable: true },
      { id: 'sustain_time', name: 'Sustain Time', type: 'float', defaultValue: 100, minValue: 10, maxValue: 500, unit: 'ms', automatable: true },
      { id: 'sensitivity', name: 'Sensitivity', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true },
      { id: 'output', name: 'Output', type: 'float', defaultValue: 0, minValue: -12, maxValue: 12, unit: 'dB', automatable: true },
      { id: 'mix', name: 'Mix', type: 'float', defaultValue: 1, minValue: 0, maxValue: 1, automatable: true },
    ],
    defaultPreset: { attack: 0, sustain: 0, attack_time: 5, sustain_time: 100, sensitivity: 0.5, output: 0, mix: 1 },
  };

export default MbTransientShaperPlugin;
